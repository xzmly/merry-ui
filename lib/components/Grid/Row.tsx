import * as React from 'react';
import {useState} from "react";
import classes from '../../helpers/classes';
import Col,{ ColProps }from "./Col"
import RowContext,{ RowContextProps }from "./Context";
import './Grid.styl';

export type SpacingType = string | number

type ChildrenType = Array<React.ReactElement<ColProps>> | React.ReactElement<ColProps>

enum Justify {
  end = "flex-end",
  start = "flex-start",
  around = "space-around",
  between = "space-between",
  center  = "center"
}
enum Align {
  start = 'flex-start',
  end = 'flex-end',
  center = 'center',
  baseline = 'baseline',
  stretch = 'stretch'
}

interface RowProps extends React.HTMLAttributes<HTMLElement> {
  type?: "block" | "flex"
  align?: "start" | "end" | "center" | "baseline" | "stretch"
  justify?: "start" | "end" | "center" | "around" | "between"
  spacing?: [SpacingType?,SpacingType?,SpacingType?,SpacingType?]
  wrap?: "wrap" | "nowrap" | "wrap-reverse"
  children: ChildrenType
}

interface RowComponent<P> extends React.FC<P> {
  Col: React.FC<ColProps>
}

const Row: RowComponent<RowProps> =
        props => {

  const [totalWidth,setTotalWidth] = useState<number>(0);

  const {
    className,
    align,
    justify,
    spacing,
    style,
    children,
    wrap,
    type,
    ...restProps
  } = props;

  const propsStyle: React.CSSProperties = {
    display: type || "block",
    flexWrap: wrap,
    alignItems: align ? Align[align] : undefined,
    justifyContent: justify ? Justify[justify] : undefined,
  };

  const ContextValue:RowContextProps = {
    totalWidth,
    grids: 24,
    spacing: spacing ? spacing.map((v:SpacingType) => v + 'px').join(' ') : undefined
  };

  return (
      <RowContext.Provider
        value={ContextValue}
      >
        <div ref={(element) => setTotalWidth(element ? element.offsetWidth : 0)}
             className={classes(className,'row')}
             style={{...style,...propsStyle}}
             {...restProps}>
          {children}
        </div>
      </RowContext.Provider>
  )
};

Row.Col = Col;

export default Row