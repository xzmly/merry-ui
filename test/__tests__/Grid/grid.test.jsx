import renderer from 'react-test-renderer'
import Row from '../../../lib/components/Grid/Row'
import React from 'react'
import {mount} from "enzyme/build/index";

const { Col } = Row;

describe('Grid',()=>{
  it('render row and col',()=>{
    const json = renderer.create(<Row>
      <Col>123</Col>
    </Row>).toJSON();
    expect(json).toMatchSnapshot()
  });
  it('spacing',()=>{
    const c = mount(
        <Row spacing={[8,8]}>
          <Col>123</Col>
        </Row>
    );
    expect(c.find('.merry-col-spacing-box').exists()).toBe(true)
  });
  it('Align Justify',()=>{
    const c = mount(
        <Row align={'center'} justify={'end'}>
          <Col style={{background: 'red'}}>123</Col>
        </Row>
    );
    expect(c.props().align).toBe('center');
    expect(c.props().justify).toBe('end');
    expect(c.find('.merry-row').getElement().props.style.alignItems).toBe('center')
    expect(c.find('.merry-row').getElement().props.style.justifyContent).toBe('flex-end')
  });
  it('col spacing pull push background',()=>{
    const c = mount(
        <Row>
          <Col style={{background: 'red'}} spacing={[8,8]} pull={5} push={10}>123</Col>
        </Row>
    );
    expect(c.find('.merry-col-spacing-box').exists()).toBe(true);
    expect(c.find('.merry-col').getElement().props.style.padding).toBe('8px 8px');
    expect(c.find('.merry-col').getElement().props.style.marginLeft).toBe('5px');
    expect(c.find('.merry-col').getElement().props.style.marginRight).toBe('10px');
  });
});