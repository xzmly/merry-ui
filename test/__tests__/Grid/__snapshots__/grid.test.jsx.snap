// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid render row and col 1`] = `
<div
  className="merry-row"
  style={
    Object {
      "alignItems": undefined,
      "display": "block",
      "flexWrap": undefined,
      "justifyContent": undefined,
    }
  }
>
  <div
    className="merry-col"
    style={
      Object {
        "background": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "padding": undefined,
        "width": undefined,
      }
    }
  >
    123
  </div>
</div>
`;
